<div style="border:solid black 2px; padding: 20px">

    
**Outlining The Task**
    
</div> 

<div style="border:solid black 2px; padding: 20px">
<div style="border:solid gray 5px; padding: 20px" class="alert alert-warning">
    
# Customers' Loyalty Program Analysis by awrb
    
</div> 

# Table of Contents: <a id='tableofcontents'></a>

    - 1. Introduction
        - 1.1 Goal
        - 1.2 Stages
    - 2. Data Pre-processing
        - 2.1 Data Overview Elaboration
        - 2.2 Rectifying Columns Names
        - 2.3 Ensuring the Data Quality
        - 2.4 Ensuring for Data Type, Array Form, Missing Values, and Duplicate Data
        - 2.5 Correcting The Missing Values and Duplicate Data
            - 2.5.1 Missing Values
            - 2.5.2 Duplicate Data
    - 3. Explorative Data Analysis (EDA)
        - 3.1 Description of Analyzing The Loyalty Program Approach
        - 3.2 Description of Calculating Average Purchase Size Method
        - 3.3 Analyzing The Loyalty Program
            - 3.3.1 Analyzing of the changes in average purchase size for each cohort for loyalty program members/non-members
            - 3.3.2 Analyzing the purchase frequency for loyalty program members/non-members
    - 4. Hypothesis Testing
        - 4.1 Formulating and Testing Statistical Hypotheses
            - 4.1.1 Testing the hypothesis that the average purchase size is higher for loyalty program members than non-members
            - 4.1.2 Formulating a statistical hypothesis regarding the data from the dataset and test it
    - 5. The Use of Metrics and Indicators Business
    - 6. Dashboard
    - 7. Presentation
    - 8. General Conclusions and Recommendations

## Introduction

        The dataset contains data on purchases made at the building-material retailer Home World. All of its customers have membership cards. Moreover, they could become members of the store's loyalty program for $20 per month. The program includes discounts, information on special offers, and gifts.

### Goal

        Therefore, we are going to take in deep in delving dataset materials in this project analysis. Accordingly, we have to define for the Project background.
        
        First, the customer is everyone who purchases at the building-material retailer Home World and they have membership cards. In this way, the project manager responsible for the loyalty program.
    
        Second, the targets of this information analysis are they who intended user of the final result. They are who the clients and retailers of the building-material Home World. Specifically the customers who initiated the projects.
        
        Third, the customers could become members of the store's loyalty program for $20 per month. The program includes discounts, information on special offers, and gifts.
        
        Fourth, this project is made to purpose for long term period or the entire period, so the development have to experiencing an upward movement. Indeed, we use the already datasets for our project making analysis.
        
        Fifth, the presentation intended for the clients and retailers. Moreover for the customers, in case by easy presentation which is easier to understand by the customers.
        
        Sixth, following the introduction of the loyalty program, it's important to understand whether it has actually led to the expected growth in average purchase size and number of items in the cart. This study will tell us how effective the program is.
        
        Seventh, this is new and fresh research. Our solution for developing the business growth with the aim of all parties benefiting each other.

### Stages

    There are several steps to complete this project, namely:

    Step 1. Downloading and Loading The Data
    Step 2. Performing Several Focus Things to be Analyzed
    Step 3. Doing Tests on The Hypothesis
    Step 4. Metrics and Indicators Business
    Step 5. Dashboard Construction
    Step 6. Preparing The Presentation
    Step 7. Formulating General Conclusions and Recommendations for Working with Customers
    
    which is divided into the following detailed stages:

    1. Data Pre-processing
    2. Explorative Data Analysis (EDA)
    3. Hypothesis Testing
    4. The Use of Metrics and Indicators Business
    5. Dashboard
    6. Presentation
    7. General Conclusions and Recommendations
    
    With the following description:

    1. Data Pre-processing (Step 1. Downloading and Loading The Data)
    - In this stage, we will load and read datasets from the following file path:

    '/datasets/retail_dataset_us.csv'
    '/datasets/product_codes_us.csv'
    
    Here are the data descriptions of the files path above:
    
    retail_dataset_us.csv contains:
    - purchaseId
    - item_ID
    - purchasedate
    - Quantity — the number of items in the purchase
    - CustomerID
    - loyalty_program — whether the customer is a member of the loyalty program
    - ShopID
    
    product_codes_us.csv contains:
    - productID
    - price_per_one
    
    2. Explorative Data Analysis (EDA) (Step 2. Performing Several Focus Things to be Analyzed)
    - In this stage, there are several things that need to be our focus, namely:
        1. Description of Analyzing The Loyalty Program Approach
        2. Description of Calculating Average Purchase Size Method
        3. Analyzing The Loyalty Program
            - Analyzing of the changes in average purchase size for each cohort for loyalty program members/non-members
            - Analyzing the purchase frequency for loyalty program members/non-members
    
    3. Hypothesis Testing (Step 3. Doing Tests on The Hypothesis)
    - In this stage, we will do the tests on the Hypothesis.
        1. Formulating and Testing Statistical Hypotheses
            - Testing the hypothesis that the average purchase size is higher for loyalty program members than non-members
            - Formulating a statistical hypothesis regarding the data from the dataset and test
    
    4. The Use of Metrics and Indicators Business (Step 4. Metrics and Indicators Business)
    - In this stage, we will do the Metrics and Indicators Business.
    
    
    5. Dashboard (Step 5. Dashboard Construction)
    - In this stage, we will create the Dashboard Construction for visualization on Tableau Public.
    
    
    6. Presentation (Step 6. Preparing The Presentation)
    - In this stage, we will prepare for the presentation of the project.
    
    
    7. General Conclusions and Recommendations (Step 7. Formulating General Conclusions and Recommendations for Working with Clients)
    - In this stage, we will draw the conclusions and formulate recommendations.

[Back to Contents](#tableofcontents)

# Analyzing The Store's Loyalty Program

## Data Pre-processing
    - Step 1. Downloading and Loading The Data

<div style="border:solid black 2px; padding: 20px">    
    
    First of all, we are going to load the required libraries of 'pandas', 'numpy', 'datetime', 'matplotlib.pyplot', 'seaborn', 'warnings', as follows:
    
</div>  

In [1]:
import pandas as pd
import numpy as np
import datetime as dt
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats as st

import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.precision', 20)
pd.set_option('display.max_columns', 100)

    Now, we will load the data files into DataFrames, as follows:

In [147]:
dataset_retail_path = pd.read_csv('/datasets/retail_dataset_us.csv', parse_dates=['purchasedate'])
codes_product_path = pd.read_csv('/datasets/product_codes_us.csv', sep = ";")

    Also upon up, we have already optimize the data for analysis purposes, columns containing dates which are converted into date data type, like the functions above.

    Now, we are going to take a look for the datas. These are the datas:

In [148]:
dataset_retail_path.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105335 entries, 0 to 105334
Data columns (total 7 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   purchaseid       105335 non-null  object        
 1   item_ID          105335 non-null  object        
 2   Quantity         105335 non-null  int64         
 3   purchasedate     105335 non-null  datetime64[ns]
 4   CustomerID       69125 non-null   float64       
 5   loyalty_program  105335 non-null  int64         
 6   ShopID           105335 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(3)
memory usage: 5.6+ MB


In [149]:
codes_product_path.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3159 entries, 0 to 3158
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   productID      3159 non-null   object 
 1   price_per_one  3159 non-null   float64
dtypes: float64(1), object(1)
memory usage: 49.5+ KB


    Findings:
    - We have two datasets, namely 'dataset_retail_path' and 'codes_product_path'.
    
    For 'dataset_retail_path' dataset:
    - The datasets amounted 105335 rows and 7 columns.
    - With data detailed, dtypes: datetime64[ns](1), float64(1), int64(2), object(3).
    - The capitalization of the columns names title still contains a mixture of upper and lowercase letters.
    - Datasets memory usage: 5.6+ MB.
    - In addition, there are some columns that comprise about customer's purchase transaction information in the table above. Those are 'purchaseid', 'item_ID', 'Quantity', 'purchasedate', 'CustomerID', 'loyalty_program', and 'ShopID'.
    

    For 'codes_product_path' dataset:
    - The datasets amounted 3159 rows and 2 columns.
    - With data detailed, dtypes: float64(1), object(1).
    - The capitalization of the columns names title still contains a mixture of upper and lowercase letters.
    - Datasets memory usage: 49.5+ KB.
    - There are two columns that comprise about product information. Those are 'productID' and 'price_per_one'.

[Back to Contents](#tableofcontents)

### Data Overview Elaboration

    In this session, we will do the data overview elaboration by using some functions of 'columns', '.info()', '.head()', and '.describe()' for checking the datas, as follows:

    For 'dataset_retail_path' dataset:

In [150]:
dataset_retail_path.columns

Index(['purchaseid', 'item_ID', 'Quantity', 'purchasedate', 'CustomerID',
       'loyalty_program', 'ShopID'],
      dtype='object')

In [151]:
dataset_retail_path.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105335 entries, 0 to 105334
Data columns (total 7 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   purchaseid       105335 non-null  object        
 1   item_ID          105335 non-null  object        
 2   Quantity         105335 non-null  int64         
 3   purchasedate     105335 non-null  datetime64[ns]
 4   CustomerID       69125 non-null   float64       
 5   loyalty_program  105335 non-null  int64         
 6   ShopID           105335 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(3)
memory usage: 5.6+ MB


In [152]:
dataset_retail_path.head()

purchaseid item_ID  Quantity        purchasedate  CustomerID  \
0     538280   21873        11 2016-12-10 12:50:00     18427.0   
1     538862   22195         0 2016-12-14 14:11:00     22389.0   
2     538855   21239         7 2016-12-14 13:50:00     22182.0   
3     543543   22271         0 2017-02-09 15:33:00     23522.0   
4     543812   79321         0 2017-02-13 14:40:00     23151.0   

   loyalty_program   ShopID  
0                0   Shop 3  
1                1   Shop 2  
2                1   Shop 3  
3                1  Shop 28  
4                1  Shop 28

In [153]:
dataset_retail_path.describe()

Quantity                  CustomerID  \
count  105335.00000000000000000000  69125.00000000000000000000   
mean        7.82121801870223531949  21019.30204701627371832728   
std       327.94669490762487384927   1765.44467926672473367944   
min    -74216.00000000000000000000  18025.00000000000000000000   
25%         0.00000000000000000000  19544.00000000000000000000   
50%         2.00000000000000000000  20990.00000000000000000000   
75%         7.00000000000000000000  22659.00000000000000000000   
max     74214.00000000000000000000  23962.00000000000000000000   

                   loyalty_program  
count  105335.00000000000000000000  
mean        0.22634451986519199540  
std         0.41846665415992301273  
min         0.00000000000000000000  
25%         0.00000000000000000000  
50%         0.00000000000000000000  
75%         0.00000000000000000000  
max         1.00000000000000000000

    For additional, we will check the range of date in the 'dataset_retail_path', as follows:

In [154]:
dataset_retail_path['purchasedate'].min(), dataset_retail_path['purchasedate'].max()

(Timestamp('2016-12-01 08:26:00'), Timestamp('2017-02-28 17:04:00'))

    For 'codes_product_path' dataset:

In [155]:
codes_product_path.columns

Index(['productID', 'price_per_one'], dtype='object')

In [156]:
codes_product_path.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3159 entries, 0 to 3158
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   productID      3159 non-null   object 
 1   price_per_one  3159 non-null   float64
dtypes: float64(1), object(1)
memory usage: 49.5+ KB


In [157]:
codes_product_path.head()

productID           price_per_one
0     10002  0.84999999999999997780
1     10080  0.84999999999999997780
2     10120  0.20999999999999999223
3    10123C  0.65000000000000002220
4    10124A  0.41999999999999998446

In [158]:
codes_product_path.describe()

price_per_one
count  3159.00000000000000000000
mean      2.95449509338398241809
std       7.21380794954317661194
min       0.00000000000000000000
25%       0.65000000000000002220
50%       1.44999999999999995559
75%       3.29000000000000003553
max     175.00000000000000000000

    Findings:
    
    - In the results above, we find that these data are in the form of datetime64[ns](1), float64(1), int64(2), object(3) from 'dataset_retail_path' dataset and in the form of float64(1), object(1) from 'codes_product_path' dataset.
    - In the result of '.head()' function, both two datasets display that all of this datas is 'almost' in the form of numbers, consisting of integers, decimal numbers, datetime, also sometimes mixed with letters.
    - The datas range are between 1 December 2016 to 28 February 2017 (Timestamp('2016-12-01 08:26:00'), Timestamp('2017-02-28 17:04:00')).

[Back to Contents](#tableofcontents)

### Rectifying Columns Names

        As one of the findings above, there is case of the capitalization of the columns names title still contains a mixture of upper and lowercase letters. We will rectify them by changing all of columns names into lower case and 'snake_case'.
        On the other hand, we will revamp the column name of 'item_id' into 'product_id' from the 'dataset_retail_path' dataset, in case of has similar values with the 'product_id' from the 'codes_product_path' dataset.

In [159]:
dataset_retail_path.columns = ['purchase_id', 'product_id', 'quantity', 'purchase_date', 'customer_id', 'loyalty_program', 
                     'shop_id']
codes_product_path.columns = ['product_id', 'price_per_one']

    Now, we look back into the table after rectifying the columns names:

    For 'dataset_retail_path' dataset:

In [160]:
dataset_retail_path.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105335 entries, 0 to 105334
Data columns (total 7 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   purchase_id      105335 non-null  object        
 1   product_id       105335 non-null  object        
 2   quantity         105335 non-null  int64         
 3   purchase_date    105335 non-null  datetime64[ns]
 4   customer_id      69125 non-null   float64       
 5   loyalty_program  105335 non-null  int64         
 6   shop_id          105335 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(3)
memory usage: 5.6+ MB


In [161]:
dataset_retail_path.head()

purchase_id product_id  quantity       purchase_date  customer_id  \
0      538280      21873        11 2016-12-10 12:50:00      18427.0   
1      538862      22195         0 2016-12-14 14:11:00      22389.0   
2      538855      21239         7 2016-12-14 13:50:00      22182.0   
3      543543      22271         0 2017-02-09 15:33:00      23522.0   
4      543812      79321         0 2017-02-13 14:40:00      23151.0   

   loyalty_program  shop_id  
0                0   Shop 3  
1                1   Shop 2  
2                1   Shop 3  
3                1  Shop 28  
4                1  Shop 28

    For 'codes_product_path' dataset:

In [162]:
codes_product_path.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3159 entries, 0 to 3158
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     3159 non-null   object 
 1   price_per_one  3159 non-null   float64
dtypes: float64(1), object(1)
memory usage: 49.5+ KB


In [163]:
codes_product_path.head()

product_id           price_per_one
0      10002  0.84999999999999997780
1      10080  0.84999999999999997780
2      10120  0.20999999999999999223
3     10123C  0.65000000000000002220
4     10124A  0.41999999999999998446

    Findings:
    
    - We have managed to rectify the capitalization title and our data contains all in lowercase letters.
    - We have done to revamp the column name of 'item_id' into 'product_id' from the 'dataset_retail_path' dataset, in case of has similar values with the 'product_id' from the 'codes_product_path' dataset.

[Back to Contents](#tableofcontents)

### Ensuring the Data Quality

    In this session, we will ensure the Data Quality both for 'dataset_retail_path' and 'codes_product_path' dataset, as follows:

    For 'dataset_retail_path' dataset:

    First, we are going to ensure the data quality of the 'purchase_id' column, as follows:

In [164]:
dataset_retail_path['purchase_id'].value_counts()

537434     675
538071     652
538349     620
537638     601
537237     597
          ... 
543310       1
C544433      1
544699       1
C544590      1
536552       1
Name: purchase_id, Length: 4894, dtype: int64

    Second, we are going to ensure the data quality of the 'product_id' (before as the: 'item_id') column, as follows:

In [165]:
dataset_retail_path['product_id'].value_counts()

85123A    565
22423     499
22469     414
22720     377
85099B    372
         ... 
20738       1
16045       1
90058B      1
84854       1
90027C      1
Name: product_id, Length: 3159, dtype: int64

    Third, we are going to ensure the data quality of the 'quantity' column, as follows:

In [166]:
sorted(dataset_retail_path['quantity'].unique())

[-74216,
 -9361,
 -2601,
 -1431,
 -1401,
 -991,
 -940,
 -742,
 -724,
 -601,
 -501,
 -401,
 -391,
 -321,
 -311,
 -301,
 -291,
 -241,
 -235,
 -231,
 -224,
 -218,
 -208,
 -207,
 -202,
 -201,
 -193,
 -178,
 -174,
 -163,
 -151,
 -146,
 -145,
 -141,
 -131,
 -121,
 -115,
 -107,
 -101,
 -100,
 -97,
 -95,
 -93,
 -89,
 -85,
 -81,
 -78,
 -77,
 -73,
 -71,
 -70,
 -65,
 -62,
 -61,
 -60,
 -58,
 -55,
 -54,
 -53,
 -51,
 -49,
 -48,
 -46,
 -45,
 -44,
 -43,
 -42,
 -41,
 -39,
 -37,
 -36,
 -34,
 -33,
 -32,
 -31,
 -30,
 -29,
 -26,
 -25,
 -24,
 -23,
 -22,
 -21,
 -20,
 -19,
 -18,
 -17,
 -16,
 -15,
 -14,
 -13,
 -12,
 -11,
 -10,
 -9,
 -8,
 -7,
 -6,
 -5,
 -4,
 -3,
 -2,
 0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,


    Temporary Finding:
    - We find datas that comprise minus values in the quantity column.
    - Now, we will find out the datas which in minus values.

    Find out the datas which in minus values in the 'quantity' column, as follows:

In [167]:
dataset_retail_path.loc[dataset_retail_path['quantity'] < 0]

purchase_id product_id  quantity       purchase_date  customer_id  \
64         C539944      22776        -2 2016-12-23 11:38:00      20239.0   
109        C542910      20726        -2 2017-02-01 15:38:00      23190.0   
112        C542426      22418       -25 2017-01-28 09:32:00      19825.0   
253        C539726      22791       -11 2016-12-21 14:24:00      22686.0   
344        C544034      21878        -2 2017-02-15 11:28:00      20380.0   
...            ...        ...       ...                 ...          ...   
105160     C541650          M        -2 2017-01-20 11:44:00          NaN   
105172     C540246      79320        -2 2017-01-05 15:43:00      18760.0   
105211     C539467      22801        -2 2016-12-19 12:46:00      20723.0   
105250     C540847      22197        -3 2017-01-11 17:35:00      19137.0   
105300     C540164      21144       -13 2017-01-05 12:02:00      20590.0   

        loyalty_program  shop_id  
64                    0  Shop 27  
109                   1  Shop 19  
112                   0  Shop 27  
253                   1   Shop 3  
344                   0   Shop 9  
...                 ...      ...  
105160                0   Shop 0  
105172                0  Shop 19  
105211                0  Shop 13  
105250                0   Shop 0  
105300                0  Shop 16  

[2118 rows x 7 columns]

    In this session, we will get the percentage of datas that comprise minus values in the 'quantity' column versus the whole data, as follows:

In [168]:
len(dataset_retail_path.loc[dataset_retail_path['quantity'] < 0]) / len(dataset_retail_path) * 100

2.01072767835952

    Temporary Findings:
    - As we could see above, there are minus values in the 'quantity' column in the initial table of 2118 rows, then we try to versus the whole data of 2%. In short, this case does not yet known what is the significance of this values or we should ask to related team. They could be the return products or something else that we still do not know. For resolve this case, we will transform the value into zero (0) values. As follows:

    Transforming the minus value in the 'quantity' column with zero.

In [169]:
dataset_retail_path.loc[dataset_retail_path['quantity'] < 0, 'quantity'] = 0
dataset_retail_path

purchase_id product_id  quantity       purchase_date  customer_id  \
0           538280      21873        11 2016-12-10 12:50:00      18427.0   
1           538862      22195         0 2016-12-14 14:11:00      22389.0   
2           538855      21239         7 2016-12-14 13:50:00      22182.0   
3           543543      22271         0 2017-02-09 15:33:00      23522.0   
4           543812      79321         0 2017-02-13 14:40:00      23151.0   
...            ...        ...       ...                 ...          ...   
105330      538566      21826         1 2016-12-13 11:21:00          NaN   
105331      540247      21742         0 2017-01-05 15:56:00      21143.0   
105332      538068      85048         1 2016-12-09 14:05:00      23657.0   
105333      538207      22818        11 2016-12-10 11:33:00      18427.0   
105334      543977      22384         9 2017-02-14 15:35:00      21294.0   

        loyalty_program  shop_id  
0                     0   Shop 3  
1                     1   Shop 2  
2                     1   Shop 3  
3                     1  Shop 28  
4                     1  Shop 28  
...                 ...      ...  
105330                0   Shop 0  
105331                0  Shop 24  
105332                1  Shop 16  
105333                0  Shop 29  
105334                0  Shop 19  

[105335 rows x 7 columns]

    Rechecking whether any the minus values after removing them, as follows:

In [170]:
dataset_retail_path.loc[dataset_retail_path['quantity'] < 0]

Empty DataFrame
Columns: [purchase_id, product_id, quantity, purchase_date, customer_id, loyalty_program, shop_id]
Index: []

    Thus, we will continue the next column.
    Fourth, we are going to ensure the data quality of the 'purchase_date' column, as follows:

In [171]:
dataset_retail_path['purchase_date'].value_counts()

2016-12-06 16:57:00    675
2016-12-09 14:09:00    652
2016-12-10 14:59:00    621
2016-12-07 15:28:00    601
2016-12-06 09:58:00    597
                      ... 
2017-02-17 17:05:00      1
2017-01-05 14:38:00      1
2017-02-21 15:11:00      1
2017-02-24 12:25:00      1
2016-12-13 17:10:00      1
Name: purchase_date, Length: 4430, dtype: int64

    Fifth, we are going to ensure the data quality of the 'customer_id' column, as follows:

In [172]:
dataset_retail_path['customer_id'].value_counts()

18427.0    796
23520.0    757
20285.0    751
20990.0    718
20590.0    549
          ... 
23956.0      1
18184.0      1
21827.0      1
20288.0      1
18182.0      1
Name: customer_id, Length: 1749, dtype: int64

    Sixth, we are going to ensure the data quality of the 'loyalty_program' column, as follows:

In [173]:
dataset_retail_path['loyalty_program'].value_counts()

0    81493
1    23842
Name: loyalty_program, dtype: int64

    And last from 'dataset_retail_path' dataset, we are going to ensure the data quality of the 'shop_id' column, as follows:

In [174]:
sorted(dataset_retail_path['shop_id'].unique())

['Shop 0',
 'Shop 1',
 'Shop 10',
 'Shop 11',
 'Shop 12',
 'Shop 13',
 'Shop 14',
 'Shop 15',
 'Shop 16',
 'Shop 17',
 'Shop 18',
 'Shop 19',
 'Shop 2',
 'Shop 20',
 'Shop 21',
 'Shop 22',
 'Shop 23',
 'Shop 24',
 'Shop 25',
 'Shop 26',
 'Shop 27',
 'Shop 28',
 'Shop 29',
 'Shop 3',
 'Shop 30',
 'Shop 4',
 'Shop 5',
 'Shop 6',
 'Shop 7',
 'Shop 8',
 'Shop 9']

    For 'codes_product_path' dataset:

    First, we are going to ensure the data quality of the 'product_id' column, as follows:

In [175]:
codes_product_path['product_id'].value_counts()

15044B    1
21682     1
90136     1
84971S    1
84723     1
         ..
22602     1
22583     1
90180A    1
37509     1
21452     1
Name: product_id, Length: 3159, dtype: int64

    Second, we are going to ensure the data quality of the 'price_per_one' column, as follows:

In [176]:
codes_product_path['price_per_one'].value_counts()

0.00000000000000000000     382
1.25000000000000000000     251
0.84999999999999997780     195
0.41999999999999998446     183
2.95000000000000017764     125
                          ... 
2.68999999999999994671       1
0.05999999999999999778       1
0.07000000000000000666       1
34.95000000000000284217      1
9.76999999999999957367       1
Name: price_per_one, Length: 192, dtype: int64

    Findings:
    
    - We have checked the Data Quality of dataset columns.
    - We found datas that comprise minus values in the 'quantity' column and now we have transformed the value into zero (0) values.
    - The data is already appropriate.

[Back to Contents](#tableofcontents)

### Ensuring for Data Type, Array Form, Missing Values, and Duplicate Data

    We will ensure the datas, as follows:

    1. 'dataset_retail_path' dataset

    Now, we ensure the Data Type for 'dataset_retail_path' dataset, as follows:

In [177]:
dataset_retail_path.dtypes

purchase_id                object
product_id                 object
quantity                    int64
purchase_date      datetime64[ns]
customer_id               float64
loyalty_program             int64
shop_id                    object
dtype: object

    We ensure the Array Form for 'dataset_retail_path' dataset, as follows:

In [178]:
dataset_retail_path.shape

(105335, 7)

    We ensure the Missing Values for 'dataset_retail_path' dataset, as follows:

In [179]:
dataset_retail_path.isna().sum()

purchase_id            0
product_id             0
quantity               0
purchase_date          0
customer_id        36210
loyalty_program        0
shop_id                0
dtype: int64

    Finally, We ensure the Duplicate Data for 'dataset_retail_path' dataset, as follows:

In [180]:
dataset_retail_path.duplicated().sum()

1049

    2. 'codes_product_path' dataset

    First, we ensure the Data Type for 'codes_product_path' dataset, as follows:

In [181]:
codes_product_path.dtypes

product_id        object
price_per_one    float64
dtype: object

    Then, we ensure the Array Form for 'codes_product_path' dataset, as follows:

In [182]:
codes_product_path.shape

(3159, 2)

    We ensure the Missing Values for 'codes_product_path' dataset, as follows:

In [183]:
codes_product_path.isna().sum()

product_id       0
price_per_one    0
dtype: int64

    Finally, we ensure the Duplicate Data for 'codes_product_path' dataset, as follows:

In [184]:
codes_product_path.duplicated().sum()

0

    Findings:
    
    - We have ensured for Data Type, Array Form, Missing Values, and Duplicate Data for both datasets.
    - The Data Type is already mentioned above, consist of datetime64[ns](1), float64(1), int64(2), object(3) for 'dataset_retail_path' dataset and float64(1), object(1) for 'codes_product_path' dataset.
    - The Array Form is (105335, 7), that means 105335 rows and 7 columns as well for 'dataset_retail_path' dataset.
    - The Array Form is (3159, 2), that means 3159 rows and 2 columns as well for 'codes_product_path' dataset.
    - We found 36210 Missing Values in the 'customer_id' column from 'dataset_retail_path' dataset.
    - We found 1049 Duplicate Data from 'dataset_retail_path' dataset.
    - Dissimilar with another dataset, for 'codes_product_path' is clean from Duplicate Data and there are no Missing Values in this dataset.

[Back to Contents](#tableofcontents)

### Correcting The Missing Values and Duplicate Data 

    Now, we will correct the Missing Values and Duplicate Data in the 'dataset_retail_path' dataset.

#### Missing Values

    Anteriorly, we will calculate the percentage of the missing values in the 'dataset_retail_path' dataset, as follows:

In [185]:
dataset_retail_path.isna().sum() / len(dataset_retail_path) * 100

purchase_id         0.0000000000000000000
product_id          0.0000000000000000000
quantity            0.0000000000000000000
purchase_date       0.0000000000000000000
customer_id        34.3760383538235174683
loyalty_program     0.0000000000000000000
shop_id             0.0000000000000000000
dtype: float64

    We will correct The Missing Values from 'customer_id' column in the 'dataset_retail_path' dataset, then saving into the new table, as follows:

In [186]:
missingvalue_customer_id = dataset_retail_path.loc[dataset_retail_path['customer_id'].isna()]
missingvalue_customer_id

purchase_id product_id  quantity       purchase_date  customer_id  \
6           540468      22189         0 2017-01-07 13:55:00          NaN   
7           541424      79321         1 2017-01-17 17:57:00          NaN   
8           538883      22891         0 2016-12-14 15:56:00          NaN   
11          541696      22197         4 2017-01-20 18:08:00          NaN   
16          539744      22300         5 2016-12-21 15:21:00          NaN   
...            ...        ...       ...                 ...          ...   
105325      536982      20719        69 2016-12-03 14:27:00          NaN   
105326      540681      22227         1 2017-01-10 16:25:00          NaN   
105327      538349      22625         0 2016-12-10 14:59:00          NaN   
105329      540154      22134         3 2017-01-05 11:30:00          NaN   
105330      538566      21826         1 2016-12-13 11:21:00          NaN   

        loyalty_program shop_id  
6                     0  Shop 0  
7                     0  Shop 0  
8                     0  Shop 0  
11                    0  Shop 0  
16                    0  Shop 0  
...                 ...     ...  
105325                0  Shop 0  
105326                0  Shop 0  
105327                0  Shop 0  
105329                0  Shop 0  
105330                0  Shop 0  

[36210 rows x 7 columns]

    Checking the distribution of missing values, as follows:

In [187]:
missingvalue_customer_id.describe()

quantity  customer_id  loyalty_program
count  36210.00000000000000000000          0.0          36210.0
mean       2.31331123998895327176          NaN              0.0
std       34.72273184722440220185          NaN              0.0
min        0.00000000000000000000          NaN              0.0
25%        0.00000000000000000000          NaN              0.0
50%        0.00000000000000000000          NaN              0.0
75%        2.00000000000000000000          NaN              0.0
max     5567.00000000000000000000          NaN              0.0

    Checking the zero value and the minus value in the 'quantity' column, as follows:

In [189]:
m_value = missingvalue_customer_id.loc[missingvalue_customer_id['quantity'] < 1]
m_value

purchase_id product_id  quantity       purchase_date  customer_id  \
6           540468      22189         0 2017-01-07 13:55:00          NaN   
8           538883      22891         0 2016-12-14 15:56:00          NaN   
23          538149      21154         0 2016-12-09 16:27:00          NaN   
28          539451      22376         0 2016-12-17 16:59:00          NaN   
32          540468      21990         0 2017-01-07 13:55:00          NaN   
...            ...        ...       ...                 ...          ...   
105297      540646     84707A         0 2017-01-10 14:32:00          NaN   
105303      537666     90214D         0 2016-12-07 18:36:00          NaN   
105311      538884      21166         0 2016-12-14 15:56:00          NaN   
105316      543013      22266         0 2017-02-02 13:35:00          NaN   
105327      538349      22625         0 2016-12-10 14:59:00          NaN   

        loyalty_program shop_id  
6                     0  Shop 0  
8                     0  Shop 0  
23                    0  Shop 0  
28                    0  Shop 0  
32                    0  Shop 0  
...                 ...     ...  
105297                0  Shop 0  
105303                0  Shop 0  
105311                0  Shop 0  
105316                0  Shop 0  
105327                0  Shop 0  

[20216 rows x 7 columns]

    Now, we calculate the percentage of the missing values in the 'customer_id' column that comprises the zero values in 'quantity' column, as follows:

In [190]:
len(m_value) / len(dataset_retail_path) * 100

19.19210139080078

    We are going to check the distribution of missing values in the 'customer_id' column that comprises the zero values, as follows:

In [192]:
m_value.describe()

quantity  customer_id  loyalty_program
count   20216.0          0.0          20216.0
mean        0.0          NaN              0.0
std         0.0          NaN              0.0
min         0.0          NaN              0.0
25%         0.0          NaN              0.0
50%         0.0          NaN              0.0
75%         0.0          NaN              0.0
max         0.0          NaN              0.0

    We will remove the missing values from the 'customer_id' column that have zero values in the 'quantity' column, as follows:

In [194]:
dataset_retail_path = dataset_retail_path.loc[~(dataset_retail_path['customer_id'].isna()) | ~(dataset_retail_path['quantity'] < 1)]

    Rechecking the rest of missing values in the 'dataset_retail_path' table, as follows:

In [195]:
dataset_retail_path.isna().sum()

purchase_id            0
product_id             0
quantity               0
purchase_date          0
customer_id        15994
loyalty_program        0
shop_id                0
dtype: int64

    Revamping the value of missing values from the 'dataset_retail_path' table with zero (0) values, as follows:

In [196]:
dataset_retail_path.loc[dataset_retail_path['customer_id'].isna(), 'customer_id'] = 0

    Rechecking the missing values in the 'dataset_retail_path' table after removing all missing values, as follows:

In [197]:
dataset_retail_path.isna().sum()

purchase_id        0
product_id         0
quantity           0
purchase_date      0
customer_id        0
loyalty_program    0
shop_id            0
dtype: int64

    Now, we have corrected the missing value in the 'dataset_retail_path' dataset.
    We are going to continue for the Duplicate Data.

    Findings:
    
    - Beforehand, we found 36210 rows Missing Values in the 'customer_id' column from 'dataset_retail_path' dataset, that means 34.37% in the percentage calculation from the whole data.
    - Afterwards, we found 20216 rows that means 19.19% the whole data which comprises missing values in the 'customer_id' column and zero values in the 'quantity' column. The prediction is there is an error system when retrieving the data.
    - For the 15994 rows missing values in the 'customer_id' column, the value has changed into 0.
    - We have done in removing all missing values. There are no missing value anymore.

[Back to Contents](#tableofcontents)

#### Duplicate Data

    Now, we will check for duplicate data in the 'dataset_retail_path' dataset.

In [201]:
dataset_retail_path.duplicated().sum()

1001

    In this moment, we find 1001 duplicate data in the 'dataset_retail_path' dataset. Then, we will remove those duplicate datas from the dataset, as follows:

In [204]:
dataset_retail_path = dataset_retail_path.drop_duplicates().reset_index(drop=True)

    Ensuring the result, as follows:

In [206]:
dataset_retail_path.duplicated().sum()

0

    Now, we have corrected the duplicate data case from 'dataset_retail_path' dataset.

    Findings:
    
    - After we fix the missing value, we found 1001 rows of duplicate data from 'dataset_retail_path' dataset.
    - And we have corrected the duplicate data. 
    - Now, the 'dataset_retail_path' dataset is clean from Duplicate Data.
    - Both datasets are ready to be analyzed and processed. 

[Back to Contents](#tableofcontents)

## Explorative Data Analysis (EDA)
    - Step 2. Performing Several Focus Things to be Analyzed

<div style="border:solid black 2px; padding: 20px">
    
                                            Carry Out Exploratory Data Analysis
    
</div>

### Description of Analyzing The Loyalty Program Approach

### Description of Calculating Average Purchase Size Method

### Analyzing The Loyalty Program

#### Analyzing of the changes in average purchase size for each cohort for loyalty program members/non-members
    - In this session, we are going to assess the changes in average purchase size for each cohort for loyalty program members/non-members.

#### Analyzing the purchase frequency for loyalty program members/non-members
    - In this session, we are going to assess the purchase frequency for loyalty program members/non-members.

[Back to Contents](#tableofcontents)

## Hypothesis Testing
    - Step 3. Doing Tests on The Hypothesis

### Formulating and Testing Statistical Hypotheses

#### Testing the hypothesis that the average purchase size is higher for loyalty program members than non-members

#### Formulating a statistical hypothesis regarding the data from the dataset and test it

[Back to Contents](#tableofcontents)

## The Use of Metrics and Indicators Business
    - Step 4. Metrics and Indicators Business

[Back to Contents](#tableofcontents)

## Dashboard
    - Step 5. Dashboard Construction

[Back to Contents](#tableofcontents)

## Presentation
    - Step 6. Preparing The Presentation

[Back to Contents](#tableofcontents)

## General Conclusions and Recommendations
    - Step 7. Formulating General Conclusions and Recommendations for Working with Clients

<div style="border:solid black 2px; padding: 20px">
    
# For Data Pre-processing

    1. Downloading, Loading and Reading The Datas
    Findings:
    - We have two datasets, namely 'dataset_retail_path' and 'codes_product_path'.

    For 'dataset_retail_path' dataset:
    - The datasets amounted 105335 rows and 7 columns.
    - With data detailed, dtypes: datetime64[ns](1), float64(1), int64(2), object(3).
    - The capitalization of the columns names title still contains a mixture of upper and lowercase letters.
    - Datasets memory usage: 5.6+ MB.
    - In addition, there are some columns that comprise about customer's purchase transaction information in the table above. Those are 'purchaseid', 'item_ID', 'Quantity', 'purchasedate', 'CustomerID', 'loyalty_program', and 'ShopID'.


    For 'codes_product_path' dataset:
    - The datasets amounted 3159 rows and 2 columns.
    - With data detailed, dtypes: float64(1), object(1).
    - The capitalization of the columns names title still contains a mixture of upper and lowercase letters.
    - Datasets memory usage: 49.5+ KB.
    - There are two columns that comprise about product information. Those are 'productID' and 'price_per_one'.
    
    2. Data Overview Elaboration
    Findings:
    - In the results above, we find that these data are in the form of datetime64[ns](1), float64(1), int64(2), object(3) from 'dataset_retail_path' dataset and in the form of float64(1), object(1) from 'codes_product_path' dataset.
    - In the result of '.head()' function, both two datasets display that all of this datas is 'almost' in the form of numbers, consisting of integers, decimal numbers, datetime, also sometimes mixed with letters.
    - The datas range are between 1 December 2016 to 28 February 2017 (Timestamp('2016-12-01 08:26:00'), Timestamp('2017-02-28 17:04:00')).
    
    3. Rectifying Columns Names
    Findings:
    - We have managed to rectify the capitalization title and our data contains all in lowercase letters.
    - We have done to revamp the column name of 'item_id' into 'product_id' from the 'dataset_retail_path' dataset, in case of has similar values with the 'product_id' from the 'codes_product_path' dataset.
    
    4. Ensuring the Data Quality
    Findings:
    - We have checked the Data Quality of dataset columns.
    - We found datas that comprise minus values in the 'quantity' column and now we have transformed the value into zero (0) values.
    - The data is already appropriate.
    
    5. Correcting The Missing Values and Duplicate Data
    a. Missing Values
        Findings:
        - Beforehand, we found 36210 rows Missing Values in the 'customer_id' column from 'dataset_retail_path' dataset, that means 34.37% in the percentage calculation from the whole data.
        - We found 20216 rows that means 19.19% the whole data which comprises missing values in the 'customer_id' column and zero values in the 'quantity' column. The prediction is there is an error system when retrieving the data.
        - For the 15994 rows missing values in the 'customer_id' column, the value has changed into 0.
        - We have done in removing all missing values. There are no missing value anymore.
    
    b. Duplicate Data
        Findings:
        - After we fix the missing value, we found 1001 rows of duplicate data from 'dataset_retail_path' dataset.
        - And we have corrected the duplicate data. 
        - Now, the 'dataset_retail_path' dataset is clean from Duplicate Data.
        - Both datasets are ready to be analyzed and processed. 

</div>

<div style="border:solid black 2px; padding: 20px">
    
# For Explorative Data Analysis (EDA)
    
</div>

<div style="border:solid black 2px; padding: 20px">

# For Hypothesis Testing
    
</div>

<div style="border:solid black 2px; padding: 20px">

# For The Use of Metrics and Indicators Business
    
</div>

<div style="border:solid black 2px; padding: 20px">

# Dashboard
    
</div>

<div style="border:solid black 2px; padding: 20px">

# Presentation
    
</div>

<div style="border:solid black 2px; padding: 20px">

# Recommendations
    
</div>

[Back to Contents](#tableofcontents)